deduce.go

```diff
@@ -597,96 +692,103 @@ func (sm *SourceMgr) deduceFromPath(path string) (deductionFuture, error) {
 	if root, err := exm.deduceRoot(path); err == nil {
 		mb, err := exm.deduceSource(path, u)
 		if err != nil {
-			return deductionFuture{}, err
+			return pathDeduction{}, err
 		}
 
-		return deductionFuture{
-			rslow: false,
-			root:  strfut(root),
-			psf:   srcfut(mb),
+		return pathDeduction{
+			root: root,
+			mb:   mb,
 		}, nil
 	}
 
-	// No luck so far. maybe it's one of them vanity imports?
-	// We have to get a little fancier for the metadata lookup by chaining the
-	// source future onto the metadata future
+	return pathDeduction{}, errNoKnownPathMatch
+}
 
-	// Declare these out here so they're available for the source future
-	var vcs string
-	var ru *url.URL
+type httpMetadataDeducer struct {
+	once       sync.Once
+	deduced    pathDeduction
+	deduceErr  error
+	basePath   string
+	returnFunc func(pathDeduction)
+	callMgr    *callManager
+}
 
-	// Kick off the vanity metadata fetch
-	var importroot string
-	var futerr error
-	c := make(chan struct{}, 1)
-	go func() {
-		defer close(c)
-		var reporoot string
-		importroot, vcs, reporoot, futerr = parseMetadata(path)
-		if futerr != nil {
-			futerr = fmt.Errorf("unable to deduce repository and source type for: %q", opath)
+func (hmd *httpMetadataDeducer) deduce(ctx context.Context, path string) (pathDeduction, error) {
+	// TODO(sdboyer) can this be replaced by the code in golang.org/x?
```

Comment:
Yeah, i tossed that comment in there as a reminder after #167. But you're right - because that system doesn't separate out scheme selection as its own thing (the `maybeSource` allow this, but also allow abstracting over vcs type for e.g. bitbucket), we can't use it.